antidote-v1.0/MP1-syringe-redesign.md

```diff
@@ -0,0 +1,171 @@
+
+Having state in an external component means we can allow everything else to be stateless.
+
+The database will PURELY be used for state management, so we can use a common source of state for all components, complete with all the necessary features like locking
+
+The API can easily be stateless because it's just designed to translate API calls into message queue messages essentially, or perform reads on teh database. It won't make direct writes to the DB. As a result, this can scale out easily.
+
+The only thing we'll have to figure out is if we need to import lesson definitions into the database. On one hand, I like the idea of pure curriculum-as-code - meaning the files on the filesystem, cloned from the Git repo, are the source of truth. However, this would mean we'd have to be make sure that we have the same version of the curriculum always in the right spot for all the microservices, which may be on different machines.
+
+On the other hand, we could do what StackStorm does with packs, which is to provide tooling to import a curriculum's "code" into database constructs. The microservices then access that data, rather than all of them accessing their own localized filesystems. The curriculum-as-code is still the original source of truth, but it's been placed into a centralized location for all to access.
+
+Maybe a "syrctl import" command is needed to facilitate this.
+
+
+
+### Design Domain - Microservices and Message Queue
+
+Services:
+- API
+- Scheduler
+- Garbage Collector
+- Stats (influxdb)
+
+NATS Streaming is the likely candidate for this. It's super simple and blazing fast.
+
+Message queue design. As long as there's a component to translate events to achievements
+then we will pick them up. If not then we will configure the messages to time out and they just go nowhere. Allows us to just plug things into the bus to enable the functionality
+
+The design allows us to enable or disable features simply by starting the relevant processes to listen on the message queue or the Syringe API. If we don't want to export to influx, we simply don't enable that process. If we don't want to enable gamification, don't start the translator process. All messages like this will be sent with a TTL, so if nothing is there to pick messages up off the queue, they'll disappear after a while.
+
+The scheduler can also scale out as long as we ensure only one of the instances of the scheduler takes a request off of the queue (i.e. no fan-out). However, the scheduler should still direct all incoming requests into a goroutine so that each individual scheduler process can handle multiple processes concurrently, just like it does today. In addition, since the scheduler will be writing to the database (as will the GC process), we need to make sure we can perform transactions on a per-UUID basis. Need to put more detailed thought into this part.
+
+### Design Area - Observability Instrumentation
+
+Are users having problems?
```

Comment:
I'm wondering : are users first class citizen entities in Antidote/Syringe ? with a profile, privileges, etc ? Which links to authentication / access control...

Would make sense only on the context of "controlled education or training" and not free to use portal as in the current installation for NRE labs

Reply:
No plans to implement access control today. Introducing any form of authentication will be opt-in, and only for the purposes of tracking progress. Not saying going further is out of the question, but IMO the simplicity of the current model is a big part of the value prop, so trying to build more robust "training" features like this shouldn't be a priority, probably for a long time.